Cleanup File & FileUtils

bcardiff opened the pull request:
* Add missing methods to File of features available in FileUtils: same_content?/cmp copy/cp
* Only `FileUtils.cp` treats the dest as a potential directory. `File.copy` works with file paths.
* I left a TODO related to #8926, #8919 for `File.copy` optimizations via syscall.
* Deprecate `Dir.rmdir` in favor of `Dir.delete` (`File.delete` is the file counter-part)
* Another breacking change is the straight removal of `FileUtils.cmp(IO, IO)` (which seems totally misplaced)
* Allow `File.same?` and `File.touch` with `Path` arguments.

Related to #5231.

The APIs that are not strict aliases in FileUtils are

* `FileUtils.touch, .mv, .mkdir, .mkdir_p, .rm`: because they allow multiple files
* `FileUtils.cp`: because it allows destination to be a directory (and allows multiple files)
* `FileUtils.cp_r`: there is no Dir counter-part and it handle the semantic of File vs Dir source / destination.
* `FileUtils.ln[_s|_sf]`: because it allows destination to be a directory and allows multiple files
* `FileUtils.rm_r[f]`: there is no Dir counter-part and allows multiple directories.

So, the following are the criteria that limited the scope of the PR are:

* I don't think we need to allow multiple input in File/Dir API.
* I'm not sure about doing the special treatment in the destination path that `cp_r`, `ln_s` will fit in File API.
* There is also some treatment of `File.symlink?` in `rm_r` that I hesistated to move into `Dir`, and so, I didn't implement `Dir.delete(recursive: true)` proposal of https://github.com/crystal-lang/crystal/issues/5231#issuecomment-343418239

bcardiff commented:
Finally I forgot to say that I would rather leave FileUtils, specially if the semantics are not 100% translatable to Dir/File since they will offer some terminal semantics I found very handy.

Side notes:
* In C# File.copy does not accept a directory as target [ref](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.copy?view=netcore-3.1)
* In Ruby File.copy was deprecated in favor of FileUtils [ref](https://apidock.com/ruby/v1_8_7_330/File/copy/class) 🙃

j8r commented on src/file_utils.cr:
```diff
@@ -480,7 +447,7 @@ module FileUtils
   #
   # NOTE: Alias of `Dir.rmdir`
   def rmdir(path : String) : Nil
```
This method can be deprecated too. I don't know how the core members sees the future of this module, but we can start now removing unnecessary aliases at least.

bcardiff replied:
I think is valid to have this unix-isms in a module. And FileUtils offers some operations that I think does not fit neatly in Dir or File.

If you say this can be deprecated because is a plain alias, then I don't agree. My intention is to keep FileUtils with similar methods and semantics as you would find in a unix shell.

I think @maiha has some opinions regarding why unix-isms are nice.

j8r replied:
I was only pointing duplicated methods. There are lots of terminal methods missing, which are very useful and very commonly used like `ls`, `head`, `tail`, `wc`, `find`, etc...

I don't think they will be accepted (will they?), because all aliases at the end can be better written in more idiomatic Crystal.

j8r replied:
Anyway, sorry it starts to be out of scope :sweat_smile: .
Let's keep it then.

maiha replied:
This is not a "plain alias" as @bcardiff says, nor is it a "duplicate method".

Since Crystal is an "object-oriented" language, the core classes such as `File` and `Dir`, their APIs are provided in an object-oriented paradigm.

On the other hand, since UNIX is a "procedural" concept, it has different APIs.

Here, the two paradigm API entry points are each defined and their **implementations** are only shared.

So, if you want to clean up, I think it's a good idea to remove the confusing annotation part of "NOTE: Alias of".

j8r replied:
Both are called in a functional way – they are class methods, not instance methods. Only the name change.

I think `FileUtils` is just there because of Ruby, a reason why I believe this is nobody asked to add `ls`, which is IMO of the most common and useful command.

`Utils` is also quite bad, if the main purpose of the module is to provide a command-like interface, it would better be called `FileCommands`or `Shell`.

j8r replied:
Both are called in a functional way – they are class methods, not instance methods. Only the name change.

I think `FileUtils` is just there because of Ruby, a reason why I believe this is nobody asked to add `ls`, which is one of the most common and useful command.

`Utils` is also quite bad, if the main purpose of the module is to provide a command-like interface, it would better be called `FileCommands`or `Shell`.

j8r replied:
I don't say everything on FileUtils has to go to trash. I'm only saying we don't have to copy blindly 1-1 the Ruby API, and review the actual module to see what is often used (`cp`, `dir` methods) and those not or confusing like `cmp` or `pwd` (I searched both on GitHub before saying this)

maiha replied:
>  nobody asked to add ls

Well, why not try to think of `FileUtils` as `File (mutable) Utils`?

asterite replied:
Members of the core team like FileUtils so it's going to stay. I know opinions are diverse here but the decision is already taken, so please let's not discuss removing anything.

j8r commented on src/file_utils.cr:
```diff
@@ -38,39 +38,10 @@ module FileUtils
-      return true if read1 == 0
-    end
+  # NOTE: Alias of `File.same_content?`
+  def cmp(filename1 : String, filename2 : String)
```
I didn't know `cmp` was a command. Same here, I also think it can be deprecated.

bcardiff replied:
I get that you want to deprecate the whole FileUtils module. But why you said this method can be deprecated specifically?

j8r replied:
`cmp` is not very common, and quite cryptic name. Not sure this alias really helps.

j8r commented on src/dir.cr:
```diff
@@ -248,7 +248,13 @@ class Dir
+  end
+
+  # Removes the directory at the given path.
+  def self.delete(path)
     Crystal::System::Dir.delete(path)
```
```suggestion
  def self.delete(path : Path | String)
    Crystal::System::Dir.delete(path.to_s)
```

j8r commented on src/file.cr:
```diff
@@ -746,7 +781,7 @@ class File < IO::FileDescriptor
   #
   # If the file does not exist, it will be created.
-  def self.touch(filename : String, time : Time = Time.utc)
+  def self.touch(filename : String | Path, time : Time = Time.utc)
```
Unrelated, the good default would be `Time.local` and not `Time.utc`, to be on par with the OS' time.
Otherwise, the file will be dated in the past/future.

straight-shoota replied:
I can't see how that would be the case. `Time.utc` and `Time.local` return the same time instance, just with different location properties.

j8r replied:
My bad, just tried it. Then, how can I touch a file with the UTC time and not my local time?

straight-shoota replied:
There's no difference because file timestamps are always in UTC.

j8r replied:
Oh right, yep. The files dates are in a UNIX epoch time - like every times in UNIX.

j8r replied:
The conversion is done by our OS.

j8r replied:
The conversion is done by the commands/tools.

bcardiff commented:
Rebased on master. And addressed @RX14 feedback.

I changed what happens on File.copy when destination exists to match the Ruby and `cp` default semantics. https://github.com/crystal-lang/crystal/pull/9175#discussion_r415261607

bcardiff commented:
@oprypin the permissions support in windows is not there yet, right? The new specs that asserts the permissions are set on the destination file in File.copy are failing.

oprypin commented:
I don't know, wasn't working on that

bcardiff commented:
Ok, thanks. 

Ignoring those specs for now. Otherwise the CI complains with:

```
Failures:

  1) File .copy copies permissions
     Failure/Error: File.info(out_path).permissions.should eq(File::Permissions.new(0o700))

       Expected: rwx------ (0o700)
            got: rw-rw-rw- (0o666)

     # \spec\std\file_spec.cr:1260

  2) File .copy overwrites existing destination and permissions
     Failure/Error: File.info(out_path).permissions.should eq(File::Permissions.new(0o700))

       Expected: rwx------ (0o700)
            got: rw-rw-rw- (0o666)

     # \spec\std\file_spec.cr:1275
```

oprypin commented:
Oh. fileutil is not tested on Windows, so makes sense that stuff is broken if you're moving it out
https://github.com/crystal-lang/crystal/blob/2cbe65b0a13c8a2ce91bd4799f5b96b8e1953097/spec/win32_std_spec.cr#L62

Sija commented:
@bcardiff Would be sensible to close #9170 perhaps, along with rest of the fixes while you're at it :)

bcardiff commented:
Done @Sija although I would say it is unrelated to the goal of this PR.

I checked Ruby code and it covers some scenarios we don't. For example, it deals with symlinks in the destination path by removing them.
